### PR TITLE
docs(resolver): Lay groundwork for documenting MSRV-aware logic

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

This is prep for document the MSRV-aware resolver (see #14639), in particular
- This give more breathing room for adding this new heuristic to the resolver documentation
- This provides the context for understanding the limitations

In moving documentation, I asked the question "where would I look to find this if I had a question on it".  I tried to balance this by not putting too much formal / technical documentation in more guide-level descriptions.  In particular, while "Specifying Dependencies" is in the reference, its also written in somewhat of a guide-style.

There is likely more work that can be done, including
- Maybe making the "SemVer Compatibility" chapter the de facto reference for Cargo's version of semver that other sections reference for a more exhaustive description.
- Splitting discussion of the Feature resolver out of the resolver and features documentation.  In the current implementation, we have 3 resolve phases (1) lockfile, (2) adapt to the current compilation, (3) resolve features.  The last two really serve the same role and I'd consider merging discussion of them.

### How should we test and review this PR?

I tried to break up changes in smaller to digest chunks.  This means some times a new section doesn't fully jive with another section until a follow up commit.  I'd recommend reviewing by commit while having the full diff up on the side to see if a concern is still relevant.

### CODE REVIEW

File: src/doc/src/reference/resolver.md
Comment (Eh2406) [inline src/doc/src/reference/resolver.md:67]: I don't understand what the sentence is trying to say.

File: src/doc/src/reference/resolver.md
Comment (Eh2406) [inline src/doc/src/reference/resolver.md:74]: Also having trouble parsing this one.

Comment (epage) [reply src/doc/src/reference/resolver.md:74]: Which part?

The first half is trying to describe our sorting of versions.

The second half is trying to deal with limitations in relying on version order, e.g. a package without an MSRV depending on one that does and we can't backtrack to pick a better version of the non-MSRV package.

File: src/doc/src/reference/resolver.md
Comment (Eh2406) [reply src/doc/src/reference/resolver.md:67]: I was confused by "Cargo would prefer to backtrack or error rather than include two versions that would have unified except their [dependency specifications] conflicted."

File: src/doc/src/reference/resolver.md
Comment (Eh2406) [reply src/doc/src/reference/resolver.md:74]: I was having trouble with "The resolver cannot backtrack because a more-preferred version becomes available."

File: src/doc/src/reference/resolver.md
Comment (epage) [reply src/doc/src/reference/resolver.md:67]: That is talking about the inability to unify `"=1.2.4"` with `"=1.5.6"`.

File: src/doc/src/reference/resolver.md
Comment (epage) [reply src/doc/src/reference/resolver.md:74]: With the explanation I gave, any thoughts on how to cover this?  Or should we skip it and leave it to the MSRV section that I'll add later?

Comment (Eh2406) [reply src/doc/src/reference/resolver.md:74]: Can you elaborate on that situation? What's the alternative design that we would love to have if it was technically possible?

The hope of these questions is that having written it out it becomes clear what point should be summarized here, or it becomes clear that it should be its own paragraph in the MSRV section. In either case the writing/thinking that is done in this comment can be used in that follow-up.

Comment (epage) [reply src/doc/src/reference/resolver.md:74]: This is less about design decisions but more about implications.  I'm trying to help the user take the principles and see the effect.

File: src/doc/src/reference/resolver.md
Comment (epage) [reply src/doc/src/reference/resolver.md:67]: > If two [dependency specifications] conflict then they can not be unified on a single version. Cargo backtracks or errors in this case.

A challenge in writing this is to be so clear we cannot be misunderstood.  saying two dependency specifications that can conflict is open ended for the user to make assumptions about what may conflict, like major versions.  This is why I was emphasizing that we attempt to unify but can't.  Working on a new sentence based on your feedback.

Comment (epage) [reply src/doc/src/reference/resolver.md:67]: Next proposal

> If multiple versions would have been unified if it wasn't for conflicts in their [dependency specifications], Cargo will backtrack, erroring if no solution is found, rather than selecting multiple versions.

File: src/doc/src/reference/resolver.md
Comment (epage) [reply src/doc/src/reference/resolver.md]: I went ahead and removed it as its hard to talk about it in a theoretical sense as its hard to describe how the resolver could hit this today.  With my MSRV documentation, it will be eaiser.

File: src/doc/src/reference/resolver.md
Comment (Eh2406) [reply src/doc/src/reference/resolver.md]: That's definitely better. We still have some ambiguity about what can and can't be unified.

**Review by Eh2406 — APPROVED** [review]
Thank you for adding this information and clarification. I think it makes the situation a lot clearer for our users!

I'm not going to single-handedly r+, because documentation requires review from many perspectives. But I'm excited to see this land once  others have had a chance to review it.

File: src/doc/src/reference/resolver.md
Comment (weihanglo) [inline src/doc/src/reference/resolver.md:106]: Should we call it default version requirement instead of caret?

File: src/doc/src/reference/manifest.md
Comment (weihanglo) [inline src/doc/src/reference/manifest.md:107]: nit: Should these link to their sections respectively?

**Review by weihanglo — APPROVED** [review]
Great work! Thanks a lot!

Left some nits but feel free to `@bors r=Eh2406,weihanglo` (not sure if it is the correct syntax for multiple reviewers)

I am unsure about how resolver actually works. Since Jacob has approved, I am comfortable :)

File: src/doc/src/reference/resolver.md
Comment (epage) [reply src/doc/src/reference/resolver.md:106]: To me, it seems weird to title something as "Default" rather than "X, which is the default".  I think this also matches for how I see people talk about version requirements.

I didn't make the version requirement syntax docs do this because

>  I tried to balance this by not putting too much formal / technical documentation in more guide-level descriptions. In particular, while "Specifying Dependencies" is in the reference, its also written in somewhat of a guide-style.

File: src/doc/src/reference/resolver.md
Comment (epage) [reply src/doc/src/reference/resolver.md:42]: The question is where to put it.  I would normally put it at the beginning of a reference but:

> I tried to balance this by not putting too much formal / technical documentation in more guide-level descriptions. In particular, while "Specifying Dependencies" is in the reference, its also written in somewhat of a guide-style.

We could have a summary section at the end that also includes the recommendations.

File: src/doc/src/reference/manifest.md
Comment (epage) [reply src/doc/src/reference/manifest.md:107]: Do you mean to their respective sections in the semver spec?

Comment (weihanglo) [reply src/doc/src/reference/manifest.md:107]: Inside this chapter just for convenience.

Linking to the official https://semver.org/ could/should happen in those sections.

Comment (epage) [reply src/doc/src/reference/manifest.md:107]: Which parts of this chapter are you referring to?  I thought this was that.

Comment (weihanglo) [reply src/doc/src/reference/manifest.md:107]: Never mind I found this is in `manifest.md`, not in `specifying-dependencies.md`. Sorry for the noise.

File: src/doc/src/reference/resolver.md
Comment (weihanglo) [reply src/doc/src/reference/resolver.md:42]: Either works for me.

Let's not block this PR on this, as the original one was in `resolver.md`, and I don't expect average users to read that chapter. Users above average don't need that either.

If someone requests, we can add it back to somewhere.

**Pull request merged**
